src/main/kotlin/org/radarbase/push/integration/garmin/service/BackfillService.kt

```
@@ -11,9 +28,86 @@ import javax.ws.rs.core.Context
+    private val requestsPerUserPerIteration: Int
+        get() = 40
+
+    private val futures: MutableList<Future<*>> = mutableListOf()
+
+    override fun onEvent(event: ApplicationEvent?) {
+        when (event?.type) {
+            INITIALIZATION_FINISHED -> start()
+            DESTROY_FINISHED -> stop()
+            else -> logger.info("Application event received: ${event?.type}")
+        }
+    }
+
+    override fun onRequest(requestEvent: RequestEvent?): RequestEventListener? = null
+
+    private fun start() {
+        logger.info("Application Initialisation completed. Starting Backfill service...")
+
+        executorService.scheduleAtFixedRate(::iterateUsers, 1, 5, TimeUnit.MINUTES)
+    }
+
+    private fun stop() {
+        logger.info("Application Destroy completed. Stopping Backfill service...")
+        try {
+            requestExecutorService.awaitTermination(30, TimeUnit.SECONDS)
+            executorService.awaitTermination(30, TimeUnit.SECONDS)
+        } catch (e: InterruptedException) {
+            logger.error("Failed to complete execution: interrupted")
+        }
+    }
+
+    private fun iterateUsers() {
+        while (futures.any { !it.isDone }) {
+            logger.info("The previous task is already running. Waiting ${WAIT_TIME_MS / 1000}s...")
+            Thread.sleep(WAIT_TIME_MS)
+        }
+        futures.clear()
+        logger.info("Making Garmin Backfill requests...")
+        try {
+            userRepository.stream().forEach { user ->
+                remoteLockManager.tryRunLocked(user.versionedId) {
+                    requestGenerator.requests(user, requestsPerUserPerIteration).forEach { req ->
+                        futures.add(requestExecutorService.submit { makeRequest(req) })
+                    }
+                }
+            }
+        } catch (exc: IOException) {
+            logger.warn("Exception while making Backfill requests.", exc)
+        }
+    }
+
+    private fun makeRequest(req: RestRequest) {
```

Please catch all exceptions from this method and log them, since otherwise the scheduler thread could be affected.

```kotlin
try {
    // make request
} catch (ex: Throwable) {
    // log something
}
```